elastic/kibana
# [Cases] Custom fields default values

Closes #171747

Fixes #174285
Fixes #174286
Fixes #174287
Fixes #174288
Fixes #174289

## Summary

This feature branch introduces default values to the custom fields in Cases.

Changes:
- The case configuration API now supports defining default values **for required custom fields**.
  - The default value field is optional.
  - The default value field must have the same type as the custom field.
- Default values for required custom fields can be configured in the cases settings page.
- The Create and Update Case pages now prefill required custom fields with their default value(if it exists).
- Create and Update case API
  - required custom field missing from the request:
    - if the default value exists -> set the custom field to the default value
    - if the default value is missing from the configuration -> throws an error
  - creating required custom fields with `value: null` -> always throws an error

### Previous PRs

#174043 #174845 #174628
#175197 #175978 #175795

## Testing

I'll leave here a couple of important use cases to test:

- Migration scenario
  1. Create **multiple** required custom fields in 8.12.
  2. Migrate to the feature branch (future 8.13).
  3. Defining default values in the settings page should work as expected.

- Update case UI
  1. Create an optional custom field
  2. Create a case and leave the custom field empty
  3. Edit the custom field to now be required
  4. Go to the case detail page and edit the custom field
  5. Is the form populated automatically with the default value?

## Release notes

Required custom fields now support default values. These default values will be used to populate automatically the custom fields if they are not defined when creating and updating cases.
